### PR TITLE
bzip2: fix hardcoded version

### PR DESCRIPTION
**Pull request by vcunat**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
On the whole, I'm a bit afraid of patching the build system this way.  I hope it would be integrated upstream soon.  Otherwise we're at risk, especially during further updates.

**Comment by K900** [comment]
Maybe we should tweak the patch so it gets the correct version from the command line instead? My m4 is pretty bad, so I'm not sure how difficult that would be.

**Comment by vcunat** [comment]
I'm no good at autotools.  (switched important stuff to meson)

**Comment by vcunat** [comment]
But my main worry is about other changes in upstream since the patch was created.

**Comment by vcunat** [comment]
[Their master branch](https://gitlab.com/bzip2/bzip2/-/tree/master) has support for cmake and meson (at a quick glance).  These might complicate our bootstrapping, but maybe we'll need either during bootstrapping sooner or later anyway.

**Comment by K900** [comment]
Suse is shipping this patch with 1.0.8 just fine, so that part should be OK (hopefully).

**Comment by vcunat** [comment]
Oh really :-)
```
Retrieving: bzip2-1.0.8-4.5.x86_64.rpm
# ls /usr/lib64/libbz2.so.1.0.*
/usr/lib64/libbz2.so.1.0.6
```

**Comment by DimStar77** [comment]
> Oh really :-)
> 
> ```
> Retrieving: bzip2-1.0.8-4.5.x86_64.rpm
> # ls /usr/lib64/libbz2.so.1.0.*
> /usr/lib64/libbz2.so.1.0.6
> ```
> 
> EDIT: I tried notifying them about this nit in tumbleweed's bzip2.

Imho, the error is assuming that library versioning is coupled in any way to package versions.

As long as the abi stays the same (no changes/additions/removals) there is no reason to change the library versioning

**Comment by vcunat** [comment]
I'm not sure what you mean by library versioning, but to me it sounds more like this part
```
BZIP2_LT_CURRENT=1
```
and corresponding SONAME-based link `libbz2.so.1 -> libbz2.so.1.0.6`

File: pkgs/tools/compression/bzip2/default.nix [hunk 1]
```diff
@@ -8,9 +8,12 @@
 # cgit) that are needed here should be included directly in Nixpkgs as
 # files.
 
+let
+  revision = "8";
+in
 stdenv.mkDerivation rec {
   pname = "bzip2";
-  version = "1.0.8";
+  version = "1.0.${revision}";
```
Comment (SuperSandro2000) [inline pkgs/tools/compression/bzip2/default.nix:16]: We don't need this

File: pkgs/tools/compression/bzip2/default.nix [hunk 2]
```diff
@@ -25,6 +28,19 @@ stdenv.mkDerivation rec {
...
+  postPatch = ''
+    patch <<-EOF
+      --- configure.ac
+      +++ configure.ac
+      @@ -3,3 +3,3 @@
+      -AC_INIT([bzip2], [1.0.6], [Julian Seward <jseward@bzip.org>])
+      +AC_INIT([bzip2], [${version}], [Julian Seward <jseward@bzip.org>])
+       BZIP2_LT_CURRENT=1
+      -BZIP2_LT_REVISION=6
+      +BZIP2_LT_REVISION=${revision}
```
Comment (SuperSandro2000) [inline pkgs/tools/compression/bzip2/default.nix:41]:
```suggestion
      +BZIP2_LT_REVISION=${lib.versions.patch version}
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Pull request merged**